symbolic-debuginfo/src/macho/bcsymbolmap.rs

```diff
@@ -187,6 +191,194 @@ impl<'a, 'd> Iterator for BcSymbolMapIterator<'a, 'd> {
 
 impl FusedIterator for BcSymbolMapIterator<'_, '_> {}
 
+#[derive(Debug, Error)]
+#[error("{kind}")]
+pub struct PListError {
+    kind: PListErrorKind,
+    #[source]
+    source: Option<Box<dyn Error + Send + Sync + 'static>>,
+}
+
+impl From<elementtree::Error> for PListError {
+    fn from(source: elementtree::Error) -> Self {
+        Self {
+            kind: PListErrorKind::Xml,
+            source: Some(Box::new(source)),
+        }
+    }
+}
+
+impl From<PListErrorKind> for PListError {
+    fn from(kind: PListErrorKind) -> Self {
+        Self { kind, source: None }
+    }
+}
+
+impl From<ParseDebugIdError> for PListError {
+    fn from(source: ParseDebugIdError) -> Self {
+        Self {
+            kind: PListErrorKind::ParseValue,
+            source: Some(Box::new(source)),
+        }
+    }
+}
+
+/// Error kind for [`PListError`].
+#[non_exhaustive]
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum PListErrorKind {
+    /// The plist did not have the expected XML schema.
+    Schema,
+    /// There was an XML parsing error.
+    Xml,
+    /// Failed to parse a PList value.
+    ParseValue,
+    /// Failed to parse UUID from filename.
+    ParseFilename,
+}
+
+impl fmt::Display for PListErrorKind {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Self::Schema => write!(f, "XML structure did not match expected schema"),
+            Self::Xml => write!(f, "Invalid XML"),
+            Self::ParseValue => write!(f, "Failed to parse a value into the right type"),
+            Self::ParseFilename => write!(f, "Failed to parse UUID from filename"),
+        }
+    }
+}
+
+/// A mapping from the `dSYM` UUID to the `BCSymbolMap` UUID.
+///
+/// When Apple compiles objects from bitcode these objects will have new UUIDs as debug
+/// identifiers.  This mapping can be found in the
+/// `dSYMs/<object-id>/Contents/Resources/<object-id>.plist` file of downloaded debugging
+/// symbols.  This struct allows you to keep track of such a mapping and provides support
+/// for parsing it from the ProperyList file format.
+#[derive(Clone, Copy, Debug)]
+pub struct BitcodeUuidMapping {
+    bitcode_uuid: DebugId,
+    dsym_uuid: DebugId,
+}
+
+impl BitcodeUuidMapping {
+    /// Parses a PropertyList containing a `DBGOriginalUUID` mapping.
+    ///
+    /// The `filename` may contain multiple path segments, the stem of the filename segment
+    /// should contain the UUID of the `dSYM`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::path::Path;
+    /// use symbolic_common::DebugId;
+    /// use symbolic_debuginfo::macho::BitcodeUuidMapping;
+    ///
+    /// let filename = Path::new("2d10c42f-591d-3265-b147-78ba0868073f.plist");
+    /// # let filename = Path::new("tests/fixtures/2d10c42f-591d-3265-b147-78ba0868073f.plist");
+    /// let dsym_uuid: DebugId = filename
+    ///     .file_stem().unwrap()
+    ///     .to_str().unwrap()
+    ///     .parse().unwrap();
```

Comment:
hm, maybe the doctests should use `?` rather than `unwrap`? So its more realistic when you want to copy-paste this verbatim

Reply:
🤷   people are meant to copy-paste?  in this case you'd use the other api provided instead of copy-pasting this anyway.  let me know if you feel strongly on this, but unless there's a strict doctest guide somewhere on which to chose i don't really mind enough on my own to change this :)

Reply:
You can have both, see [Examples on `File`](https://doc.rust-lang.org/stable/std/fs/struct.File.html#examples). 

Doc examples are primarily meant to be illustrative and idiomatic examples of how to use this in real scenarios. Since `.unwrap()` is a code smell, we should also not be documenting it. It only feels so natural because we overuse `unwrap` in unit tests to get nice stack traces.

Reply:
oh yes, doc examples are certainly idiomatic: https://doc.rust-lang.org/stable/std/primitive.str.html#examples-46

in all seriousness, yes we should aim to do better (and this already is!) but this is awkward as we have multiple error types to get rid of and we don't have anyhow in the dependencies.  i don't really mind unwrap calls in examples, the examples are not about teaching error handling...

Reply:
not to mention the more infrastructure you pile into the hidden part of the doc example the harder it becomes to maintain, especially once the nesting no longer matches.